File: app/src/main/java/fr/legrand/oss117soundboard/presentation/ui/reply/ReplyListFragment.java
package fr.legrand.oss117soundboard.presentation.ui.reply;

import android.graphics.ColorMatrix;
import android.graphics.ColorMatrixColorFilter;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import fr.legrand.oss117soundboard.R;
import fr.legrand.oss117soundboard.data.entity.Reply;
import fr.legrand.oss117soundboard.presentation.ui.reply.ui.ReplyListAdapter;
import fr.legrand.oss117soundboard.presentation.presenter.ReplyListPresenter;
import fr.legrand.oss117soundboard.presentation.ui.base.BaseFragment;
import fr.legrand.oss117soundboard.presentation.ui.listener.OnSearchListener;
import fr.legrand.oss117soundboard.presentation.ui.reply.ui.ReplyListViewHolder;
import fr.legrand.oss117soundboard.presentation.ui.view.viewinterface.ReplyListView;
import fr.legrand.oss117soundboard.presentation.ui.reply.item.ReplyViewData;

/**
 * Created by Benjamin on 30/09/2017.
 */

public class ReplyListFragment extends BaseFragment implements ReplyListView, ReplyListViewHolder.OnReplyListenListener, OnSearchListener {

    private static final String FAVORITE_KEY = "fr.legrand.oss117soundboard.presentation.ui.reply.ReplyListFragment.FAVORITE_KEY";

    @Inject
    ReplyListPresenter replyListPresenter;
    @Inject
    ReplyListAdapter replyListAdapter;

    @BindView(R.id.fragment_reply_list_recycler)
    RecyclerView searchRecycler;
    @BindView(R.id.fragment_reply_list_placeholder)
    RelativeLayout searchPlaceholder;
    @BindView(R.id.fragment_reply_list_refresh_layout)
    SwipeRefreshLayout refreshLayout;
    @BindView(R.id.fragment_reply_list_placeholder_image)
    ImageView searchPlaceholderImage;
    @BindView(R.id.fragment_reply_list_placeholder_text)
    TextView searchPlaceholderText;


    public static ReplyListFragment newInstance(boolean fromFavorite) {
        Bundle args = new Bundle();

        args.putBoolean(FAVORITE_KEY, fromFavorite);

        ReplyListFragment fragment = new ReplyListFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
        View homeView = LayoutInflater.from(getActivity()).inflate(R.layout.fragment_reply_list, container, false);
        getFragmentComponent().inject(this);
        ButterKnife.bind(this, homeView);

        replyListPresenter.setReplyListView(this);

        setPlaceholderToGrayScale();

        return homeView;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initializeRecyclerView();
        replyListPresenter.getAllReply(getArguments().getBoolean(FAVORITE_KEY));
    }

    @Override
    public void updateSearchReplyList(List<ReplyViewData> replyViewDataList) {
        replyListAdapter.setItems(replyViewDataList);
    }

    @Override
    public void displayPlaceholder() {
        searchRecycler.setVisibility(View.GONE);
        searchPlaceholder.setVisibility(View.VISIBLE);
        refreshLayout.setRefreshing(false);
    }

    @Override
    public void displayReplyList() {
        searchRecycler.setVisibility(View.VISIBLE);
        searchPlaceholder.setVisibility(View.GONE);
        refreshLayout.setRefreshing(false);
    }

    @Override
    public void updateFavoriteReply(ReplyViewData replyViewData) {
        replyListAdapter.updateItem(replyViewData, getArguments().getBoolean(FAVORITE_KEY));
        if (replyListAdapter.getItemCount() == 0) {
            displayPlaceholder();
        } else {
            displayReplyList();
        }
    }

    @Override
    public void onReplyListenClick(int position) {
        replyListPresenter.incrementReplyCount(replyListAdapter.getItem(position).getReply().getId());
    }

    @Override
    public void onReplyFavoriteClick(int position) {
        Reply reply = replyListAdapter.getItem(position).getReply();
        replyListPresenter.updateFavoriteReply(reply.getId(), !reply.isFavorite());
    }

    @Override
    public void onSearch(String search) {
        boolean fromFavorite = getArguments().getBoolean(FAVORITE_KEY);
        if (search == null || search.isEmpty()) {
            replyListPresenter.getAllReply(fromFavorite);
        } else {
            replyListPresenter.getAllReplyWithSearch(search, fromFavorite);
        }
    }

    private void initializeRecyclerView() {
        refreshLayout.setColorSchemeResources(R.color.colorPrimary);
        refreshLayout.setOnRefreshListener(() -> replyListPresenter.requestRefreshData());
        searchRecycler.setLayoutManager(new LinearLayoutManager(getActivity()));
        searchRecycler.setAdapter(replyListAdapter);
        replyListAdapter.setOnReplyListenListener(this);
    }

    private void setPlaceholderToGrayScale() {
        ColorMatrix matrix = new ColorMatrix();
        matrix.setSaturation(0);
        ColorMatrixColorFilter cf = new ColorMatrixColorFilter(matrix);
        searchPlaceholderImage.setColorFilter(cf);
        searchPlaceholderImage.setImageAlpha(128);
    }


}

File: app/src/main/java/fr/legrand/oss117soundboard/presentation/ui/reply/ReplyListFragment.kt
package fr.legrand.oss117soundboard.presentation.ui.reply

import android.arch.lifecycle.ViewModelProviders
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.view.View
import fr.legrand.oss117soundboard.R
import fr.legrand.oss117soundboard.presentation.extensions.hide
import fr.legrand.oss117soundboard.presentation.extensions.observeSafe
import fr.legrand.oss117soundboard.presentation.extensions.setToGrayScale
import fr.legrand.oss117soundboard.presentation.extensions.show
import fr.legrand.oss117soundboard.presentation.ui.base.BaseVMFragment
import fr.legrand.oss117soundboard.presentation.ui.main.ReplySharedViewModel
import fr.legrand.oss117soundboard.presentation.ui.reply.ui.ReplyListAdapter
import kotlinx.android.synthetic.main.fragment_reply_list.*
import javax.inject.Inject

/**
 * Created by Benjamin on 30/09/2017.
 */

private const val FAVORITE_KEY = "FAVORITE_KEY"

class ReplyListFragment : BaseVMFragment<ReplyListViewModel>() {
    override val viewModelClass = ReplyListViewModel::class
    @Inject
    lateinit var replyListAdapter: ReplyListAdapter

    private lateinit var sharedViewModel: ReplySharedViewModel

    override fun getLayoutId() = R.layout.fragment_reply_list

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        sharedViewModel = ViewModelProviders.of(activity!!, viewModelFactory)[ReplySharedViewModel::class.java]

        viewModel.favorite = arguments?.getBoolean(FAVORITE_KEY) ?: false

        initializeRecyclerView()

        fragment_reply_list_placeholder_image.setToGrayScale()

        sharedViewModel.searchRequested.observeSafe(this) {
            viewModel.updateSearch(it)
            viewModel.getAllReply()
        }
        sharedViewModel.replyUpdated.observeSafe(this) {
            viewModel.getAllReply()
        }

        viewModel.replyUpdated.observeSafe(this) {
            sharedViewModel.onReplyUpdated()
        }

        viewModel.replyListLiveData.observeSafe(this) {
            replyListAdapter.setItems(it)
        }

        viewModel.viewState.observeSafe(this) {
            if (it.displayingPlaceholder) {
                displayPlaceholder()
            } else {
                displayReplyList()
            }
            fragment_reply_list_refresh_layout.isRefreshing = it.refreshing
        }
    }

    private fun displayPlaceholder() {
        fragment_reply_list_recycler.hide()
        fragment_reply_list_placeholder.show()
    }

    private fun displayReplyList() {
        fragment_reply_list_recycler.show()
        fragment_reply_list_placeholder.hide()
    }


    private fun initializeRecyclerView() {
        fragment_reply_list_refresh_layout.setColorSchemeResources(R.color.colorPrimary)
        fragment_reply_list_refresh_layout.setOnRefreshListener {
            viewModel.getAllReply()
        }
        fragment_reply_list_recycler.layoutManager = LinearLayoutManager(activity)
        fragment_reply_list_recycler.adapter = replyListAdapter

        replyListAdapter.onListenClickListener = {
            viewModel.incrementReplyCount(it)
        }
        replyListAdapter.onFavoriteClickListener = { id, favorite ->
            viewModel.updateFavoriteReply(id, favorite)
        }
    }

    companion object {
        fun newInstance(fromFavorite: Boolean): ReplyListFragment {
            val args = Bundle().apply { putBoolean(FAVORITE_KEY, fromFavorite) }
            return ReplyListFragment().apply { arguments = args }
        }
    }


}